File: lettercount.rb
# 入力した英文の中で、重複した文字の数が最も多い単語を抽出する
# 条件１）重複文字数が複数の単語で同数の場合、より先頭に近い単語を抽出
# 条件２）重複が無い場合は、「-1」を出力
def LetterCountI(str)
  # デフォルト値
  count = 0
  word = "-1"
  repeated_defalt_count = 0

  # 単語毎に分割する
  word_array = str.split(" ")

  word_array.each do |words|
    # 単語を各文字に分割する
    letters_array = words.split("")
    # 重複している文字のみ抽出
    repeated = letters_array.select{|a| letters_array.index(a)!=letters_array.rindex(a)}.uniq
    # 重複した文字をカウント
    repeated_count = repeated.length
    # 重複文字が、デフォルト<抽出した数の場合、以下を実行
    if repeated_defalt_count < repeated_count
      countNew = 0
      while repeated_defalt_count < repeated_count
        repeated_count -= 1
        countNew += 1
      end
      if countNew > count
        count = countNew
        word = words
      end
    end
  end

  puts "条件１）文字が最も重複した単語を出力（最大重複が同じ場合、より先頭の単語）"
  puts "条件２）重複した文字を含む単語がない場合は「-1」"
  puts word
end

puts "英文を入力して下さい"
puts LetterCountI(STDIN.gets)

File: lettercount.js
function LetterCountI(str) {

  // to LowerCase：呼び出す文字列を小文字に変換
  str.toLowerCase();

  var arr = str.split(" ");

  var count = 0;
  var word = "-1";

  // for文「while文」、++x：加算
  // i = 単語の数、つまりsplitした単語の数だけ以下の処理を実行する
  for (var i = 0; i < arr.length; i++) {
    // a = 単語の文字数、つまり単語内の文字の数だけ以下の処理を実行する
    for (var a = 0; a < arr[i].length; a++) {
      var countNew = 0; // カウントの再定義
      // b = 重複した文字、つまり単語内の文字aに対して重複があるだけ以下の処理を実行する
      for (var b = a + 1; b < arr[i].length; b++) {
        // 重複を発見次第、countNewにカウント
        if(arr[i][a] === arr[i][b])
            countNew += 1;
      }
      // 上記のfor文を繰り返し、countNewの値が0より大きい場合、countに定義。左記の処理を行った単語をwordに定義
      if (countNew > count) {
        count = countNew;
        word = arr[i];
      }
      // wordの値を返す
      return word;
    }
  }
}
